File: scala/75.scala
object Solution {
    def sortColors(nums: Array[Int]): Unit = {
        def swap(i: Int, j: Int) {
            val tmp = nums(i)
            nums(i) = nums(j)
            nums(j) = tmp
        }
        def sort(i: Int, l: Int, r: Int) {
            if (i <= r) 
              nums(i) match {
                  case 0 => { swap(l, i); sort(i+1, l+1, r) }
                  case 2 => { swap(i, r); sort(i, l, r-1) }
                  case _ => sort(i+1, l, r)
              }
        }       
        sort(0, 0, nums.length-1)
    }
}

File: leetcodeJava/75.java
package leetcodeJava;

class Solution {
    private void swap(int[] nums, int i, int j) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }
    public void sortColors(int[] nums) {
        int l = 0, r = nums.length - 1;
        for (var i = 0; i <= r; i++) {
            if (nums[i] == 2) {
                swap(nums, i, r);
                r--;
                if (nums[i] == 2) {
                    i--;
                    continue;
                }
            }
            if (i >= 0 && nums[i] == 0) {
                swap(nums, i, l);
                l++;
            }
        }
    }
}
